nav2_costmap_2d/src/costmap_subscriber.cpp

```diff
@@ -0,0 +1,81 @@
+// Copyright (c) 2019 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <string>
+
+#include "nav2_costmap_2d/costmap_subscriber.hpp"
+
+namespace nav2_costmap_2d
+{
+
+CostmapSubscriber::CostmapSubscriber(
+  rclcpp::Node::SharedPtr ros_node,
+  std::string & topic_name)
+: node_(ros_node),
+  topic_name_(topic_name),
+  costmap_received_(false),
+  costmap_(nullptr)
+{
+  costmap_sub_ = node_->create_subscription<nav2_msgs::msg::Costmap>(topic_name,
+      std::bind(&CostmapSubscriber::costmap_callback, this, std::placeholders::_1));
+}
+
+Costmap2D * CostmapSubscriber::getCostmap()
```

Comment:
Instead of providing the costmap to the clients, can we define an interface that provides what the clients really need, i.e. poseWithinMap(), poseCost(), etc.

Reply:
I think the costmap_subscriber would eventually become the client:

![client](https://user-images.githubusercontent.com/39749557/52814434-04a64300-3051-11e9-93e1-1acbcc4e7840.png)

Reply:
Perhaps, though I would address this as a future feature of the world model in a different PR. Also, I'm not sure how much client-specific context we'd really like at the interface level. Calling something like `poseCost()` is very context-specific and I'm not sure in the scope of the WorldModel. In my view, the WorldModel could be a plugin container for abstract map representations (of which costmap2D is one) of the environment from which clients can grab and use as needed on their end. We can continue this discussion